app/models/course/achievement.rb

```diff
@@ -8,9 +8,13 @@ class Course::Achievement < ActiveRecord::Base
   belongs_to :course, inverse_of: :achievements
   has_many :course_user_achievements, class_name: Course::UserAchievement.name,
                                       inverse_of: :achievement, dependent: :destroy
-  has_many :course_users, through: :course_user_achievements, class_name: CourseUser.name
   has_many :achievement_conditions, class_name: Course::Condition::Achievement.name,
                                     inverse_of: :achievement, dependent: :destroy
+  # Due to the through relationship, destroy dependent had to be added for course users in order for
+  # UserAchievement's destroy callbacks to be called, However, this destroy dependent will not
+  # actually remove the course users when the Achievement object is destroyed.
+  has_many :course_users, through: :course_user_achievements, class_name: CourseUser.name,
+                          dependent: :destroy
```

Comment:
@custode From my test, the `course_user_achivements` get destroyed  after I removed the `dependent: :destroy` here.

Can you add a test which will fail without this ?